### ISSUE TITLE
scylla might enter infinite loop in write_fragmented

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#8401]
Seen e.g. in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/782/artifact/logs-all.release.2/1617406220040_cql_additional_tests.TestCQL.filter_by_counter_test/node1.log
With scylla version f186de909d461003e83ce24f256c2fb6884b6004
```
Reactor stalled for 524 ms on shard 1. Backtrace: 0x3f8de34 0x3f8d0b1 0x3f8dcff 0x7faf603271df 0x7faf5f72f3af 0x1b60cb0 0x2273a68 0x2273f68 0x1d4a476 0x2df6ac6 0x2df5791 0x2df4ca7 0x2df73c0 0x3fa0cdf 0x3fa1ec7 0x3fc0518 0x3f6c59a 0x93f8 0x101902
Reactor stalled for 987 ms on shard 1. Backtrace: 0x3f8de34 0x3f8d0b1 0x3f8dcff 0x7faf603271df 0x1b60c85 0x2273a68 0x2273f68 0x1d4a476 0x2df6ac6 0x2df5791 0x2df4ca7 0x2df73c0 0x3fa0cdf 0x3fa1ec7 0x3fc0518 0x3f6c59a 0x93f8 0x101902
Reactor stalled for 1852 ms on shard 1. Backtrace: 0x3f8de34 0x3f8d0b1 0x3f8dcff 0x7faf603271df 0x1b60d1f 0x2273a68 0x2273f68 0x1d4a476 0x2df6ac6 0x2df5791 0x2df4ca7 0x2df73c0 0x3fa0cdf 0x3fa1ec7 0x3fc0518 0x3f6c59a 0x93f8 0x101902
Reactor stalled for 3522 ms on shard 1. Backtrace: 0x3f8de34 0x3f8d0b1 0x3f8dcff 0x7faf603271df 0x43f76ef
Reactor stalled for 6801 ms on shard 1. Backtrace: 0x3f8de34 0x3f8d0b1 0x3f8dcff 0x7faf603271df 0x43f76ef
INFO  2021-04-02 23:28:11,972 [shard 0] compaction_manager - Asked to stop
INFO  2021-04-02 23:28:11,972 [shard 0] compaction_manager - Stopping 0 ongoing compactions
INFO  2021-04-02 23:28:11,972 [shard 0] init - Signal received; shutting down
INFO  2021-04-02 23:28:11,972 [shard 0] init - Shutting down redis service
INFO  2021-04-02 23:28:11,972 [shard 0] init - Shutting down redis service was successful
INFO  2021-04-02 23:28:11,972 [shard 0] init - Shutting down view builder
INFO  2021-04-02 23:28:11,972 [shard 0] view - Stopping view builder
```

```
gdb -c node1-scylla.556758.1617406219.core scylla/libexec/scylla

(gdb) bt
#0  0x0000000001b60ca9 in write_fragmented<managed_bytes_basic_view<(mutable_view)1>, ser::buffer_view<bytes_ostream::fragment_iterator> > (dest=..., src=...) at ./utils/fragment_range.hh:306
#1  managed_bytes::managed_bytes<ser::buffer_view<bytes_ostream::fragment_iterator> > (this=0x7faf5d1d2d98, v=...)
    at ./utils/managed_bytes.hh:486
#2  cql3::expr::(anonymous namespace)::next_value (iter=..., cdef=<optimized out>) at cql3/expr/expression.cc:365
#3  cql3::expr::(anonymous namespace)::get_non_pk_values (selection=..., static_row=..., row=<optimized out>)
    at cql3/expr/expression.cc:386
#4  cql3::expr::is_satisfied_by (restr=..., partition_key=..., clustering_key=..., static_row=..., row=<optimized out>, 
    row@entry=0x7faf5d1d3400, selection=..., options=...) at cql3/expr/expression.cc:635
#5  0x0000000002273a69 in cql3::selection::result_set_builder::restrictions_filter::do_filter (this=<optimized out>, 
    selection=..., partition_key=..., clustering_key=..., static_row=..., row=0x7faf5d1d3400) at cql3/selection/selection.cc:439
#6  0x0000000002273f69 in cql3::selection::result_set_builder::restrictions_filter::operator() (this=0x7faf5d1d36b0, 
    selection=..., partition_key=..., clustering_key=..., static_row=..., row=0x601004ec1330) at cql3/selection/selection.cc:495
#7  0x0000000001d4a477 in cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::restrictions_filter>::accept_new_row (this=0x7faf5d1d3660, static_row=..., row=...) at ./cql3/selection/selection.hh:369
#8  0x0000000002df6ac7 in cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::restrictions_filter>::accept_new_row (this=0x7faf5d1d3660, key=..., static_row=..., row=...) at ./cql3/selection/selection.hh:363
#9  service::pager::query_pager::query_result_visitor<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::restrictions_filter> >::accept_new_row (this=0x7faf5d1d3660, key=..., static_row=..., row=...)
    at service/pager/query_pagers.cc:301
#10 query::result_view::consume<service::pager::query_pager::query_result_visitor<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::restrictions_filter> >&> (this=<optimized out>, this@entry=0x7faf5d1d3770, slice=..., visitor=...) at ./query-result-reader.hh:188
#11 0x0000000002df5792 in service::pager::query_pager::handle_result<cql3::selection::result_set_builder::visitor<cql3::selection::result_set_builder::restrictions_filter> > (this=this@entry=0x601004ecbc00, visitor=..., results=..., page_size=5000, now=...) at service/pager/query_pagers.cc:335
#12 0x0000000002df4ca8 in service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}::operator()(service::storage_proxy::coordinator_query_result) const (this=<optimized out>, qr=...) at service/pager/query_pagers.cc:258
#13 0x0000000002df73c1 in std::__invoke_impl<void, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result>(std::__invoke_other, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result&&) (__f=..., __args=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:60
#14 std::__invoke<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result>(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result&&) (__fn=..., __args=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:95
#15 std::invoke<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result>(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result&&) (__fn=..., __args=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/functional:88
#16 seastar::internal::future_invoke<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result>(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, service::storage_proxy::coordinator_query_result&&) (func=..., v=...) at ./seastar/include/seastar/core/future.hh:1211
#17 seastar::future<service::storage_proxy::coordinator_query_result>::then_impl_nrvo<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<void> >(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&)#1}::operator()(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&) const::{lambda()#1}::operator()() const (this=<optimized out>) at ./seastar/include/seastar/core/future.hh:1582
#18 seastar::futurize<void>::satisfy_with_result_of<seastar::future<service::storage_proxy::coordinator_query_result>::then_impl_nrvo<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<void> >(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&)#1}::operator()(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&) const::{lambda()#1}>(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&) (pr=..., func=...) at ./seastar/include/seastar/core/future.hh:2117
#19 seastar::future<service::storage_proxy::coordinator_query_result>::then_impl_nrvo<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<void> >(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&)#1}::operator()(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&) const (this=0x601004eccf00, pr=..., func=..., state=...) at ./seastar/include/seastar/core/future.hh:1575
#20 seastar::continuation<seastar::internal::promise_base_with_type<void>, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<service::storage_proxy::coordinator_query_result>::then_impl_nrvo<service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}, seastar::future<void> >(service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, service::pager::filtering_query_pager::fetch_page(cql3::selection::result_set_builder&, unsigned int, std::chrono::time_point<gc_clock, std::chrono::duration<long, std::ratio<1l, 1l> > >, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >)::{lambda(service::storage_proxy::coordinator_query_result)#1}&, seastar::future_state<service::storage_proxy::coordinator_query_result>&&)#1}, service::storage_proxy::coordinator_query_result>::run_and_dispose() (this=0x601004eccf00) at ./seastar/include/seastar/core/future.hh:767
#21 0x0000000003fa0ce0 in seastar::reactor::run_tasks (this=0x6010000a6000, tq=...) at ./seastar/src/core/reactor.cc:2263
#22 seastar::reactor::run_some_tasks (this=this@entry=0x6010000a6000) at ./seastar/src/core/reactor.cc:2672
#23 0x0000000003fa1ec8 in seastar::reactor::run (this=0x6010000a6000) at ./seastar/src/core/reactor.cc:2831
#24 0x0000000003fc0519 in seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_97::operator()() const (this=0x600000233e00) at ./seastar/src/core/reactor.cc:4022
#25 std::__invoke_impl<void, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_97&>(std::__invoke_other, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_97&) (__f=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:60
#26 std::__invoke_r<void, seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_97&>(seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_97&) (__fn=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:110
#27 std::_Function_handler<void (), seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::$_97>::_M_invoke(std::_Any_data const&) (__functor=...) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:291
#28 0x0000000003f6c59b in std::function<void ()>::operator()() const (this=0x7faf5d1d2d98) at /usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:622
#29 seastar::posix_thread::start_routine (arg=0x7faf5d1d2d98) at ./seastar/src/core/posix.cc:60
#30 0x00007faf6031c3f9 in start_thread () from /tmp/scylla/libreloc/libpthread.so.0
#31 0x00007faf5f6cb903 in clone () from /tmp/scylla/libreloc/libc.so.6

(gdb) frame

  #0  0x0000000001b60ca9 in write_fragmented<managed_bytes_basic_view<(mutable_view)1>, ser::buffer_view<bytes_ostream::fragment_iterator> > (dest=..., src=...)
    at ./utils/fragment_range.hh:306
306             memcpy(dest.current_fragment().data(), src.current_fragment().data(), n);
(gdb) p n
$11 = 0
(gdb) p src
$12 = {
  _first = {
    _M_len = 0,
    _M_str = 0x601000158000 ""
  },
  _total_size = 8,
  _next = {
    _current = 0x6010002ed000
  }
}
(gdb) p *src._next._current
$13 = {
  next = {
    _M_t = {
      <std::__uniq_ptr_impl<bytes_ostream::chunk, std::default_delete<bytes_ostream::chunk> >> = {
        _M_t = {
          <std::_Tuple_impl<0, bytes_ostream::chunk*, std::default_delete<bytes_ostream::chunk> >> = {
            <std::_Tuple_impl<1, std::default_delete<bytes_ostream::chunk> >> = {
              <std::_Head_base<1, std::default_delete<bytes_ostream::chunk>, true>> = {
                <std::default_delete<bytes_ostream::chunk>> = {<No data fields>}, <No data fields>}, <No data fields>}, 
            <std::_Head_base<0, bytes_ostream::chunk*, false>> = {
              _M_head_impl = 0x0
            }, <No data fields>}, <No data fields>}
      }, <No data fields>}
  },
  offset = 183,
  size = 976,
  data = 0x6010002ed010 ""
}
(gdb) p dest
$14 = <optimized out>
```

This looks related / caused by 4739df2cb171bac39fb6b8c21d02e6984d67afac (Merge 'cql3: remove linearizations in the write path' from @michoecho)

So `src.current_fragment()` is empty.
There is no particular protection in `buffer_view::remove_current` against having an empty fragment at `_first` while `_total_size > 0`, but I'm not sure if that's supposed to be possible or not.  If so, we could skip it.

**Comment by haaawk** [comment]
Isn't this a duplicate of a problem @michoecho has already found and fixed on Friday? https://github.com/scylladb/scylla/issues/8398 -> https://github.com/scylladb/scylla/pull/8397

**Comment by bhalevy** [comment]
> Isn't this a duplicate of a problem @michoecho has already found and fixed on Friday? #8398 -> #8397

Looks so

**Comment by bhalevy** [comment]
Duplicate of ##8398
